### ISSUE TITLE
Backend file move error

### ISSUE THREAD

**Issue by Vec7or** [contao/contao#410]
**Affected version(s)**
Contao-Version: 4.7.2
PHP-Version: 7.2.16

**Description**  
When moving files in the file-manager inside the back-end, the move is not detected correctly.
Instead of detecting a file move a removal is detected as shown in Screenshot01.
![Screenshot01](https://user-images.githubusercontent.com/43302112/55082137-6a024080-50a1-11e9-8c42-9d245181e604.png)
Front-end articles and sites which reference the moved files are not adjusted correctly which results in file not found exceptions.
If instead of moving files inside the back-end file manager the files are moved by the `mv` command on the server console, the move is detected correctly as shown in Screenshot02.
![Screenshot02](https://user-images.githubusercontent.com/43302112/55082139-6a024080-50a1-11e9-98f4-3a61274cb007.png)

**How to reproduce**  
1. Move a file in the file-manager and run the synchronization. (Only deletion is detected)
2. Move a file on the command-line with `mv` command. (Move is detected correctly)

**Expected behaviour**
The move should also be detected correctly when done in the file-manager.

**Comment by joehi** [comment]
**Affected version(s)**
Contao-Version: 4.7.2
PHP-Version: 7.2.16

**Description**
When moving folders with files or individual files inside the BE the file paths will not be updated in the database. Luckily the wrong path stays at least in the database, so by moving the files back to the original folder will undo the effect and the paths are valid again.
However, renaming a folder within the BE is detected and updated correctly in the database.
The checkbox "Nicht synchronisieren" (do not synchronize) in the folder properties is unset. But even when I set it there was no difference in behaviour, besides the fact that then also the renaming of the foldername leads to non-updated paths in the database.

Doing the same with Contao-Version 4.4.36 works as expected and the paths in the database get updated correctly.

[Meldung im Contao Forum dazu](https://community.contao.org/de/showthread.php?74527-Bei-Verschieben-von-Dateien-wird-Pfad-nicht-in-Datenbank-upgedated&p=500431&viewfull=1#post500431)

**Comment by ausi** [comment]
Caused by b546be78e90c5289c5ce317f330b3d0821defa87  
`Dbafs::shouldBeSynchronized()` returns `false` for non-existent files since that change.

Can be fixed by changing https://github.com/contao/contao/blob/2e98cb8a2458731d474b190d5bde9cc891e1f7b8/core-bundle/src/Resources/contao/drivers/DC_Folder.php#L679-L686 to:

```diff
{
+       $syncSource = $this->blnIsDbAssisted && Dbafs::shouldBeSynchronized($source);
        $this->Files->rename($source, $destination);

        // Update the database AFTER the file has been moved
        if ($this->blnIsDbAssisted)
        {
-               $syncSource = Dbafs::shouldBeSynchronized($source);
                $syncTarget = Dbafs::shouldBeSynchronized($destination);
```

But we should probably instead fix `Dbafs::shouldBeSynchronized()` I guess. This `return true;` causes the issue: https://github.com/contao/contao/commit/b546be78e90c5289c5ce317f330b3d0821defa87#diff-917ba09b282cfda0fdd06b9133b01f74R829

**Comment by leofeyer** [comment]
@ausi Is this what you had in mind?

```php
protected static function isFileSyncExclude($strPath)
{
	if (Config::get('uploadPath') == 'templates')
	{
		return true;
	}

	$rootDir = System::getContainer()->getParameter('kernel.project_dir');

	// If $strPath is not a directory, either because the path points to a file
	// or because it does not exist at all (see #410), apply dirname() and let
	// Folder::isUnsynchronized() check if there is an unsynced parent folder
	if (!is_dir($rootDir . '/' . $strPath))
	{
		$strPath = \dirname($strPath);
	}

	$uploadPath = System::getContainer()->getParameter('contao.upload_path');

	// Outside the files directory
	if (strncmp($strPath . '/', $uploadPath . '/', \strlen($uploadPath) + 1) !== 0)
	{
		return true;
	}

	return (new Folder($strPath))->isUnsynchronized();
}
```

**Comment by ausi** [comment]
`new Folder($strPath)` creates the folder if it doesn’t exist which is not what we want in this case I think. We should probably use a while loop with `dirname()` until `is_dir()` is true and then call `isUnsynchronized()` on that path.

**Comment by leofeyer** [comment]
You are right.

```php
protected static function isFileSyncExclude($strPath)
{
	if (Config::get('uploadPath') == 'templates')
	{
		return true;
	}

	$rootDir = System::getContainer()->getParameter('kernel.project_dir');

	// Look for an existing parent folder (see #410)
	if (!file_exists($rootDir . '/' . $strPath))
	{
		do
		{
			$strCurrent = \dirname($strPath);
		}
		while ($strCurrent != '.' && !is_dir($rootDir . '/' . $strCurrent));

		if ($strCurrent != '.')
		{
			$strPath = $strCurrent;
		}
	}

	if (is_file($rootDir . '/' . $strPath))
	{
		$strPath = \dirname($strPath);
	}

	$uploadPath = System::getContainer()->getParameter('contao.upload_path');

	// Outside the files directory
	if (strncmp($strPath . '/', $uploadPath . '/', \strlen($uploadPath) + 1) !== 0)
	{
		return true;
	}

	return (new Folder($strPath))->isUnsynchronized();
}
```

**Comment by ausi** [comment]
Wouldn’t this be simpler?

```php
protected static function isFileSyncExclude($strPath)
{
	if (Config::get('uploadPath') == 'templates')
	{
		return true;
	}

	$rootDir = System::getContainer()->getParameter('kernel.project_dir');

	// Look for an existing parent folder (see #410)
	while ($strPath != '.' && !is_dir($rootDir . '/' . $strPath))
	{
		$strPath = \dirname($strPath);
	}
	
	if ($strPath == '.') 
	{
		return true;
	}

	$uploadPath = System::getContainer()->getParameter('contao.upload_path');

	// Outside the files directory
	if (strncmp($strPath . '/', $uploadPath . '/', \strlen($uploadPath) + 1) !== 0)
	{
		return true;
	}

	return (new Folder($strPath))->isUnsynchronized();
}
```

**Comment by leofeyer** [comment]
Fixed in fb6db76a11ba0f402ea138aaaca5a9a01877feb6.
